### PR TITLE
Keep track of formation assignments

### PR DESCRIPTION
**Pull request by StanislavStefanov**
**Description**

Changes proposed in this pull request:
- Create formationAssignments when assigning to formation
- Delete formationAssignments when unassigning from formation

**Pull Request status**
- [x] Implementation
- [x] Unit tests
- [x] Integration tests
- [x] `chart/compass/values.yaml` is updated 
- [x] Mocks are regenerated, using the automated script

### CODE REVIEW

**Review by PetarTodorovv — CHANGES_REQUESTED** [review]

File: components/director/internal/domain/formationassignment/service.go [hunk 1]
```diff
@@ -126,7 +175,17 @@ func (s *service) ListByFormationIDs(ctx context.Context, formationIDs []string,
 	return s.repo.ListByFormationIDs(ctx, tnt, formationIDs, pageSize, cursor)
 }
 
-// Update updates a formation assignment matching ID `id` using `in`
+// ListFormationAssignmentsForObject retrieves all Formation Assignment objects for formation with ID `formationID` that have `objectID` as source or target
+func (s *service) ListFormationAssignmentsForObject(ctx context.Context, formationID, objectID string) ([]*model.FormationAssignment, error) {
```
[the remark this answers is not in the document]
Comment (StanislavStefanov) [reply components/director/internal/domain/formationassignment/service.go:179]: We are not listing the assignments by their id rather we are looking for assignment which have their source property set to ObjectID or their target property set to ObjectID. For example if we have an application that is part of the formation we want to find the assignments in which the application takes part. So we need to find all of the assignments for which the application is source and all of the assignments for which the application is target.

File: components/director/internal/domain/formationassignment/service.go [hunk 2]
```diff
@@ -178,3 +237,224 @@ func (s *service) Exists(ctx context.Context, id string) (bool, error) {
...
+	log.C(ctx).Infof("Finished processing %d formation assignments", len(formationAssignmentsForObject)+1)
+
+	return tx.Commit()
+}
+
+// CreateOrUpdateFormationAssignment prepares the `State` and `Config` of the formation assignment based on the response and saves it in the database
+func (s *service) CreateOrUpdateFormationAssignment(ctx context.Context, assignment *model.FormationAssignment, response *webhookdir.Response) error {
+	if response == nil || *response.ActualStatusCode == *response.SuccessStatusCode {
+		assignment.State = string(model.ReadyAssignmentState)
+	}
```
[the remark this answers is not in the document]
Comment (StanislavStefanov) [reply components/director/internal/domain/formationassignment/service.go:332]: This part is refactored in the next PR. There we pass the notification and it is clear whether we have notification or no.

File: components/director/internal/domain/formationassignment/service.go [hunk 3]
```diff
@@ -178,3 +237,224 @@ func (s *service) Exists(ctx context.Context, id string) (bool, error) {
...
+
+	if response != nil && response.IncompleteStatusCode != nil && *response.ActualStatusCode == *response.IncompleteStatusCode {
```
[the remark this answers is not in the document]
Comment (StanislavStefanov) [reply components/director/internal/domain/formationassignment/service.go:334]: This is already refactored in the next PR

File: components/director/internal/domain/formationassignment/service.go [hunk 4]
```diff
@@ -178,3 +237,224 @@ func (s *service) Exists(ctx context.Context, id string) (bool, error) {
...
+	if response != nil && response.Error != nil && *response.Error != "" {
+		assignment.State = string(model.CreateErrorAssignmentState)
+		marshaled, err := json.Marshal(struct{ Error string }{Error: *response.Error})
+		if err != nil {
+			return errors.Wrapf(err, "While preparing error message from response for assignment with ID %q", assignment.ID)
+		}
+		assignment.Value = marshaled
+	}
+
+	if _, err := s.Create(ctx, s.formationAssignmentConverter.ToInput(assignment)); err != nil {
```
[the remark this answers is not in the document]
Comment (StanislavStefanov) [reply components/director/internal/domain/formationassignment/service.go:351]: It was preparation for the nex PR. We have already adapted it there.

File: components/director/internal/repo/list.go [hunk 5]
```diff
@@ -105,6 +120,11 @@ func (l *universalLister) List(ctx context.Context, resourceType resource.Type,
 	return l.listWithTenantScope(ctx, resourceType, tenant, dest, NoLock, additionalConditions)
 }
 
+// ListConditionTree lists tenant scoped entities with tenant isolation subquery based on equal condition on tenantColumn.
```
[the remark this answers is not in the document]
Comment (StanislavStefanov) [reply components/director/internal/repo/list.go:123]: The doc may be a bit missleading. This method only works for entities with embeded tenant. So the tenant isolation is done by equal condition for the tenant column. Will remove the word subquery from the doc

File: components/director/internal/domain/formationassignment/repository.go [hunk 6]
```diff
@@ -154,7 +151,25 @@ func (r *repository) ListByFormationIDs(ctx context.Context, tenantID string, fo
 	return faPages, nil
 }
 
-// Update updates the formation assignment matching the ID of the input model
+// ListAllForObject retrieves all FormationAssignment objects for formation with ID `formationID`that have objectID as `target` or `source` from the database that are visible for `tenant`
+func (r *repository) ListAllForObject(ctx context.Context, tenant, formationID, objectID string) ([]*model.FormationAssignment, error) {
+	var entitiesWithSourceObjectID EntityCollection
+	var entitiesWithTargetObjectID EntityCollection
```
[the remark this answers is not in the document]
Comment (StanislavStefanov) [reply components/director/internal/domain/formationassignment/repository.go:157]: It is leftover. We have already removed it in the next PR

File: components/director/internal/domain/formationassignment/service.go [hunk 7]
```diff
@@ -178,3 +237,224 @@ func (s *service) Exists(ctx context.Context, id string) (bool, error) {
...
+
+	if response != nil && response.IncompleteStatusCode != nil && *response.ActualStatusCode == *response.IncompleteStatusCode {
+		assignment.State = string(model.ConfigPendingAssignmentState)
+	}
+
+	if response != nil && response.Config != nil && *response.Config != "" {
+		assignment.Value = []byte(*response.Config)
+	}
+
+	if response != nil && response.Error != nil && *response.Error != "" {
```
[the remark this answers is not in the document]
Comment (StanislavStefanov) [reply components/director/internal/domain/formationassignment/service.go:342]: Such cases are handled by the webhook client. If the response status code does not match successStatusCode or incompleteStatus code the webhook client will return error

**Review by PetarTodorovv — APPROVED** [review]

File: components/director/internal/model/formation_assignment.go [hunk 8]
```diff
@@ -37,6 +37,22 @@ type FormationAssignmentPage struct {
 	TotalCount int
 }
 
+// FormationAssignmentState represents the possible states a formation assignment can be in
+type FormationAssignmentState string
+
+const (
+	// InitialAssignmentState indicates that nothing has been done with the formation assignment
+	InitialAssignmentState FormationAssignmentState = "INITIAL"
```
Comment (dzahariev) [inline components/director/internal/model/formation_assignment.go:45]: Ir will be easier for maintenance if the constant names use the following naming schema:
<Object><State>
In this case names to be:
AssignmentStateInitial
AssignmentStateReady
AssignmentStateConfigPending
...
As when you want to reference the object state in code will use autocomplete to look for the object and will see the states that are applicable for this object.

**Review by dzahariev — APPROVED** [review]
I had usability related comment that is not a blocker for merge.

File: components/director/internal/domain/runtime_context/repository.go [hunk 9]
```diff
@@ -304,6 +304,42 @@ func (r *pgRepository) ListByScenariosAndRuntimeIDs(ctx context.Context, tenant
...
+	if err != nil {
+		return nil, apperrors.NewInvalidDataError("tenantID is not UUID")
+	}
+
+	var entities RuntimeContextCollection
+
+	// Scenarios query part
+	scenariosFilters := make([]*labelfilter.LabelFilter, 0, len(scenarios))
+	for _, scenarioValue := range scenarios {
+		query := fmt.Sprintf(`$[*] ? (@ == "%s")`, scenarioValue)
```
Comment (alextargov) [inline components/director/internal/domain/runtime_context/repository.go:322]: We can extract this common query that is being used in a couple of places. It's a suggestion, not mandatory.

**Review by desislavaa — APPROVED** [review]

**Review by PetarTodorovv — APPROVED** [review]

**Pull request merged**